### DISCUSSION TITLE
Define scopes using authWithOAuth2Code

### DISCUSSION THREAD

**Discussion by dimitarnikolovv** [pocketbase/pocketbase#4326] [Q&A] [answered]
Hi, I am wondering if it is possible to define Google oauth scopes when using `authWithOAuth2Code`. I am handling the user creation in an `onRecordBeforeAuthWithOAuth2Request` hook and I am wondering if I can access things such as the user's gender and birth date in the `oAuth2User` object there. 

But to do this I need to define 'https://www.googleapis.com/auth/userinfo.email' and 'https://www.googleapis.com/auth/userinfo.profile' scopes.

**Comment by ganigeorgiev** [comment]
`authWithOAuth2Code` doesn't deal with scopes.

The scopes are specified when you send request to the provider auth page, aka. before the redirect. Or in other words, you can modify directly the auth provider url with the scopes you need (scopes are defined as `scope` query parameter; multiple scopes need to be separated by space).

**Reply by ganigeorgiev** [reply]
Also note that both `userinfo.profile` and `userinfo.email` are the default scopes for the Google provider if nothing else is specified so you don't need to specify them explicitly if that are the scopes you want.

**Reply by dimitarnikolovv** [reply]
I made a mistake I meant `user.gender.read` and `user.birthday.read`. But I got it, thanks.

**Reply by dimitarnikolovv** [reply]
Figured it out, but still not sure how to access the data in the pocketbase hook using the provided `oAuth2User` object

**Reply by ganigeorgiev** [reply]
The raw provider auth response should be under the [`rawUser` field](https://pocketbase.io/jsvm/interfaces/auth.AuthUser.html#rawUser).

**Reply by dimitarnikolovv** [reply]
Yes, I assumed the same, but the data that should come from the `user.gender.read` and `user.birthday.read` is missing.

Here is what I get from the `rawUser` filed:

```json
{...
"rawUser": {
      "verified_email": true,
      "name": "Example Example",
      "given_name": "Example",
      "family_name": "Example",
      "picture": "https://lh3.googleusercontent.com/a/ACg8ocexample_ceiAxmN7S3ZKcm_e33_pDljqW4=s96-c",
      "locale": "bg",
      "id": "000000000000000000000",
      "email": "example@gmail.com"
    }
}
```

**Reply by ganigeorgiev** [reply, marked as answer]
You may have to send a separate request to another endpoint to retrieve the data you need as I'm not sure whether `https://www.googleapis.com/oauth2/v1/userinfo` will return the fields in the response.

**Reply by ganigeorgiev** [reply]
_I also just noticed that there is a v2 api so I'll consider updating to it with the next major release._

**Reply by dimitarnikolovv** [reply]
So if I understand you correctly I should do something like this in the hooks:

```js
  const res = $http.send({
    url: 'https://people.googleapis.com/v1/people/me?personFields="names,emailAddresses"',
    method: 'GET',
    headers: {
      Authorization: 'Bearer <access-token>',
    },
  });

```

**Reply by dimitarnikolovv** [reply]
Made it work. It is not perfect, but It gets the job done. 

```js
const res = $http.send({
    url: `https://people.googleapis.com/v1/people/${
      e.oAuth2User.id
    }?personFields=${encodeURIComponent('birthdays,genders')}`,
    method: 'GET',
    headers: {
      Authorization: 'Bearer ' + e.oAuth2User.accessToken,
    },
  });
```

That's inside the pb hook.

**Comment by dimitarnikolovv** [comment]
If someone is wondering the same here is an example, correct me if it could be better. 

```js
const authMethods = await locals.pb.collection('users').listAuthMethods();

const redirectURL = `${url.origin}/api/oauth2`;
    
const googleAuth = authMethods.authProviders[0];

const authProviderRedirect = new URL(`${googleAuth.authUrl}${redirectURL}`);

authProviderRedirect.searchParams.set(
      'scope',
      'https://www.googleapis.com/auth/user.birthday.read https://www.googleapis.com/auth/user.gender.read https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/userinfo.profile'
    );

const state = googleAuth.state;
const verifier = googleAuth.codeVerifier;

cookies.set('state', state);
cookies.set('verifier', verifier);

throw redirect(302, authProviderRedirect);

```

And in the hook: 

```js
const res = $http.send({
    url: `https://people.googleapis.com/v1/people/${
      e.oAuth2User.id
    }?personFields=${encodeURIComponent('birthdays,genders')}`,
    method: 'GET',
    headers: {
      Authorization: 'Bearer ' + e.oAuth2User.accessToken,
    },
  });
```

**Reply by blujedis** [reply]
@dimitarnikolovv is this still working for you? When it hits my Oauth response route it's giving me an error in the PocketBase logs saying it doesn't have credentials. I get the appropriate consent screen but it bombs when it hits the response route.

**Reply by dimitarnikolovv** [reply]
Yes, it is still working, but the project is locked on PB 0.21.2. I haven't worked with the newer versions, so I am not sure if this example is still valid.

**Reply by blujedis** [reply]
LOL just figured it out. I was missing a scope. Basically specifying these scopes overrides what is sent by default I'm assuming as I was missing the "userinfo.profile" 

Once added it worked.
